app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTypeTelemetry.kt

```diff
@@ -0,0 +1,83 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import androidx.annotation.VisibleForTesting
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.OnLifecycleEvent
+import androidx.lifecycle.ProcessLifecycleOwner
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.COLD
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.WARM
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.HOT
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.COLD_SAVED_INSTANCE
+import org.mozilla.fenix.components.metrics.Event.AppStartupType.Type.WARM_SAVED_INSTANCE
+import org.mozilla.fenix.HomeActivity
+
+/**
+ * Tracks the startup type for application through [Event.AppStartupType]. Consider app to be
+ * "opened" only if application was previously backgrounded or application didn't exist.
+ */
+class AppStartupTypeTelemetry(private val metrics: MetricController) : LifecycleObserver {
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    // keeps track of whether we already recorded the metric
+    private var isMetricRecorded = false
+
+    // keeps track of whether the onCreate() was called for FenixApplication.
+    // helps choose which startup type to report during onHomeActivityOnCreate()
+    private var applicationOnCreateCalled = false
+
+    fun onFenixApplicationOnCreate() {
+        applicationOnCreateCalled = true
+    }
+
+    /**
+     * When onCreate for [HomeActivity] is called, we have 4 possible types of startup.
+     * this functions contains the logic of reporting the correct startup type.
+     */
+    fun onHomeActivityOnCreate(hasSavedInstanceState: Boolean) {
+        if (hasSavedInstanceState && applicationOnCreateCalled) {
+            recordAppStartupType(COLD_SAVED_INSTANCE)
+        } else if (applicationOnCreateCalled) {
+            recordAppStartupType(COLD)
+        } else if (hasSavedInstanceState) {
+            recordAppStartupType(WARM_SAVED_INSTANCE)
+        } else {
+            recordAppStartupType(WARM)
+        }
+
+        applicationOnCreateCalled = false
```

Comment:
This line of code feels misleading because of the variable name – if the application onCreate was called, how can it be uncalled? I'd suggest `wasApplicationOnCreateJustCalled` or even more specific, `wasApplicationOnCreateCalledBeforeHomeActivityOnCreate` (though it's getting a little lengthy - maybe you can shorten with `wasAppCreateCalledBeforeActivityCreate`?). With a name like that, it's very easy to understanding what a value represents and doesn't, validate that it's implemented correctly, and to read through the class quickly without second guessing your understanding.